notifications/postgres/notificator.go

```diff
@@ -0,0 +1,336 @@
+		n.isRunningMutex.Lock()
+		defer n.isRunningMutex.Unlock()
+		n.isRunning = isRunning
+		if !isRunning {
+			log.C(n.ctx).WithError(err).Info("connection to db closed, closing all consumers")
+			n.closeAllConsumers()
+		}
+	})
+	n.setConnection(connection)
+	return nil
+}
+
+type notifyEventPayload struct {
+	PlatformID     string `json:"platform_id"`
+	NotificationID string `json:"notification_id"`
+	Revision       int64  `json:"revision"`
+}
+
+func (n *notificator) updateLastKnownRevision(revision int64) {
+	n.revisionMutex.Lock()
+	defer n.revisionMutex.Unlock()
+	n.lastKnownRevision = revision
+}
+
+func (n *notificator) processNotifications(notificationChannel <-chan *pq.Notification) {
+	for pqNotification := range notificationChannel {
+		if pqNotification == nil {
+			continue
+		}
+		payload, err := n.getPayload(pqNotification.Extra)
+		if err != nil {
+			log.C(n.ctx).WithError(err).Error("could not unmarshal notification payload")
+			n.closeAllConsumers() // Ensures no notifications are lost
+		} else {
+			n.updateLastKnownRevision(payload.Revision)
+			n.processNotificationPayload(payload)
+		}
+	}
+}
+
+func (n *notificator) getPayload(data string) (*notifyEventPayload, error) {
+	payload := &notifyEventPayload{}
+	if err := json.Unmarshal([]byte(data), payload); err != nil {
+		return nil, err
+	}
+	return payload, nil
+}
+
+func (n *notificator) processNotificationPayload(payload *notifyEventPayload) {
+	notificationPlatformID := payload.PlatformID
+	notificationID := payload.NotificationID
+	recipients := n.getRecipients(notificationPlatformID)
+	if len(recipients) == 0 {
+		return
+	}
+	notification, err := n.storage.GetNotification(n.ctx, notificationID)
+	if err != nil {
+		log.C(n.ctx).WithError(err).Errorf("notification %s could not be retrieved from the DB, closing consumers", notificationID)
+		n.closeAllConsumers()
+		return
+	}
+	for _, platformConsumers := range recipients {
+		n.sendNotificationToPlatformConsumers(platformConsumers, notification)
+	}
+}
+
+func (n *notificator) getRecipients(platformID string) consumers {
+	n.consumersMutex.RLock()
+	defer n.consumersMutex.RUnlock()
+	if platformID == "" {
+		return n.consumers
+	}
+	platformConsumers, found := n.consumers[platformID]
+	if !found {
+		return nil
+	}
+	return consumers{
+		platformID: platformConsumers,
+	}
+}
+
+func (n *notificator) sendNotificationToPlatformConsumers(platformConsumers []notifications.NotificationQueue, notification *types.Notification) {
+	for _, consumer := range platformConsumers {
+		if err := consumer.Enqueue(notification); err != nil {
+			log.C(n.ctx).WithError(err).Infof("consumer %s notification queue returned error %v", consumer.ID(), err)
+			consumer.Close()
+		}
+	}
+}
+
+func (n *notificator) awaitTermination() {
+	<-n.ctx.Done()
+	logger := log.C(n.ctx)
+	logger.Info("context cancelled, stopping notificator...")
+	n.isRunning = false
+	n.stopConnection()
+	n.group.Done()
```

Comment:
It's a little bit confusing when adding 1 to a waitgroup is in one method and calling.Done is in another place. A suggestion - not sure if it makes sense - we could spawn goroutines with a utility method that takes care of both adding 1 and decrementing 1 from the waitgroup.

```
func StartInWaitGroup(f func(), g *sync.WaitGroup) {
	g.Add(1)
	go func() {
		defer g.Done()
		f()
	}()
}
```